projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/debug/LLVMDebugObjectImpl.java

```diff
@@ -0,0 +1,239 @@
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.nodes.intrinsics.llvm.debug;
+
+import com.oracle.truffle.llvm.runtime.debug.LLVMDebugBasicType;
+import com.oracle.truffle.llvm.runtime.debug.LLVMDebugDecoratorType;
+import com.oracle.truffle.llvm.runtime.debug.LLVMDebugObject;
+import com.oracle.truffle.llvm.runtime.debug.LLVMDebugType;
+import com.oracle.truffle.llvm.runtime.debug.LLVMDebugValueProvider;
+import com.oracle.truffle.llvm.runtime.floating.LLVM80BitFloat;
+
+import java.util.Map;
+import java.util.Objects;
+
+abstract class LLVMDebugObjectImpl implements LLVMDebugObject {
+
+    private static final Object[] NO_KEYS = new Object[0];
+
+    protected final long offset;
+
+    protected final LLVMDebugValueProvider value;
+
+    protected final LLVMDebugType type;
+
+    LLVMDebugObjectImpl(LLVMDebugValueProvider value, long offset, LLVMDebugType type) {
+        this.value = value;
+        this.offset = offset;
+        this.type = type;
+    }
+
+    abstract Object getValue();
+
+    @Override
+    public Object getType() {
+        return type;
+    }
+
+    @Override
+    public String toString() {
+        return Objects.toString(getValue());
+    }
+
+    static final class Enum extends LLVMDebugObjectImpl {
+
+        Enum(LLVMDebugValueProvider value, long offset, LLVMDebugType type) {
+            super(value, offset, type);
+        }
+
+        @Override
+        Object getValue() {
+            if (!value.canReadId(offset, (int) type.getSize())) {
+                return "<unknown>";
+            }
+            final long id = value.readId(offset, (int) type.getSize());
+            final Object val = type.getElementName(id);
+            return val != null ? val : "<unknown>";
+        }
+
+        @Override
+        public Object[] getKeys() {
+            return NO_KEYS;
+        }
+
+        @Override
+        public Object getMember(Object identifier) {
+            return null;
+        }
+    }
+
+    static final class Primitive extends LLVMDebugObjectImpl {
+
+        Primitive(LLVMDebugValueProvider value, long offset, LLVMDebugType type) {
+            super(value, offset, type);
+        }
+
+        @Override
+        public Object[] getKeys() {
+            return NO_KEYS;
+        }
+
+        @Override
+        public Object getMember(Object identifier) {
+            return null;
+        }
+
+        @Override
+        public Object getValue() {
```

Comment:
we might want to hide getValue (and probably all other methods of LLVMDebugObjectImpl) behind TruffleBoundaries